src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.regexp;
+
+import static org.truffleruby.language.RubyGuards.isRubyRegexp;
+
+import org.joni.Regex;
+import org.truffleruby.Layouts;
+import org.truffleruby.builtins.CoreClass;
+import org.truffleruby.builtins.CoreMethod;
+import org.truffleruby.builtins.CoreMethodArrayArgumentsNode;
+import org.truffleruby.core.regexp.RegexpNodes.ToSNode;
+import org.truffleruby.core.rope.CodeRange;
+import org.truffleruby.core.rope.Rope;
+import org.truffleruby.core.string.StringNodes.StringAppendPrimitiveNode;
+import org.truffleruby.core.string.StringOperations;
+import org.truffleruby.language.dispatch.CallDispatchHeadNode;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.DynamicObjectFactory;
+
+@CoreClass("Truffle::RegexpOperations")
+public class TruffleRegexpNodes {
+
+    @CoreMethod(names = "union", onSingleton = true, required = 2, rest = true)
+    public static abstract class RegexpUnionNode extends CoreMethodArrayArgumentsNode {
+
+        @Child StringAppendPrimitiveNode appendNode = StringAppendPrimitiveNode.create();
+        @Child ToSNode toSNode = ToSNode.create();
+        @Child CallDispatchHeadNode copyNode = CallDispatchHeadNode.create();
+
+        @Specialization(guards = "argsMatch(cachedArgs, args)", limit = "getCacheLimit()")
+        public Object executeFastUnion(VirtualFrame frame, DynamicObject str, DynamicObject sep, Object[] args,
+                @Cached(value = "args", dimensions = 1) Object[] cachedArgs,
+                @Cached("buildUnion(frame, str, sep, args)") DynamicObject union) {
+            return copyNode.call(frame, union, "clone");
```

Comment:
Rather than clone, it seems nicer to create the Regexp object here, and only cache the Regex, options and encodingCache. That way, the cache does not hold to a user-level object, and there is no extra object kept alive.
But it does mean 3 `@Cached` or some BuildRegexpData triple to hold the data needed by the constructor.

Reply:
This is probably not worth the effort.